### PR TITLE
unix: sanity-check fds before closing

### PR DESCRIPTION
**Pull request by bnoordhuis**
RFC from @piscisaureus and @saghul.

The idea is to catch a pretty common type of error that we have in node.js, namely closing stdio file descriptors.

It also ties in with joyent/node#6271 where libuv triggers a Linux kernel bug that results in a busy loop.

### CODE REVIEW

File: test/test-stdio-over-pipes.c [hunk 1]
```diff
@@ -43,7 +43,6 @@
 
 
 static void close_cb(uv_handle_t* handle) {
-  printf("close_cb\n");
```
Comment (bnoordhuis) [inline test/test-stdio-over-pipes.c:46]: This was one of those bugs: calling printf() after fd 1 is closed.

File: src/unix/udp.c [hunk 2]
```diff
@@ -40,8 +40,11 @@
 void uv__udp_close(uv_udp_t* handle) {
   uv__io_close(handle->loop, &handle->io_watcher);
   uv__handle_stop(handle);
-  close(handle->io_watcher.fd);
-  handle->io_watcher.fd = -1;
+
+  if (handle->io_watcher.fd != -1) {
```
Comment (saghul) [inline src/unix/udp.c:44]: I guess that before the change if you just did uv_udp_init and then close, fd stayed at -1 but we didn't care about the error, right?

Comment (bnoordhuis) [inline src/unix/udp.c:44]: Yep.  It did a `close(-1)` which isn't harmful but kind of sloppy (shows up in strace as failing with EBADF.)

File: src/unix/core.c [hunk 3]
```diff
@@ -406,6 +406,26 @@ int uv__accept(int sockfd) {
...
 
 
+int uv__close(int fd) {
+  int saved_errno;
+  int rc;
+
+  assert(fd > -1);  /* Catch uninitialized io_watcher.fd bugs. */
+  assert(fd > STDERR_FILENO);  /* Catch stdio close bugs. */
+
+  saved_errno = errno;
```
Comment (saghul) [inline src/unix/core.c:416]: Before this we didn't save the errno (only in some cases) so is this now a Good Thing (TM) ?

**Comment by saghul** [comment]
LGTM :+1:

File: src/unix/core.c (hunk 3 above)
Comment (bnoordhuis) [inline src/unix/core.c:416]: Early versions of uv-unix saved errno in a lot of places but I kind of scaled back on that.  For uv__close(), it's mostly a convenience thing - it often gets called from error paths where the original errno needs to be preserved.  Dealing with that in uv__close() makes that easier.

Comment (saghul) [inline src/unix/core.c:416]: I see. Thanks for explaining!

**Comment by bnoordhuis** [comment]
Thanks @saghul, landed in 359d667.

**Pull request closed without merging**